### PR TITLE
Feature/erc 223 support

### PR DESCRIPTION
**Pull request by AntoineRondelet**
This PR adds the support of all ERC20 compatible tokens (included ERC223 tokens)
See: #22

### CODE REVIEW

**Comment by AntoineRondelet** [comment]
Just need to refactor the code a bit. I'll reopen it later

**Comment by AntoineRondelet** [comment]
Move all the `Deposit` and `Withdraw` logics into their own functions `DepositLogic` and `WithdrawLogic`.
The goal here was to avoid duplicate code. 

Moreover, the call to the ERC20 compatible token is made via: `UntrustedErc20Token`, in order to illustrate the fact that we are doing a call to an external (and not trusted) contract (see: https://consensys.github.io/smart-contract-best-practices/recommendations/#mark-untrusted-contracts for further details).

Finally, in order to prevent any security breach to be used in case of malicious codes, the external calls are moved to the very end of each functions (see: https://consensys.github.io/smart-contract-best-practices/known_attacks/#reentrancy).

In order to do this, we introduced a variable `entrySaved` which is a `memory` variable (see: https://github.com/clearmatics/mobius/blob/feature/ERC-223-support/contracts/Mixer.sol#L282), This variable, in returned to the calling function via a `return` statement and aims to return a copy on the `entry` accessed in the `storage` (see: `Data storage entry = m_rings[ring_id];` in: https://github.com/clearmatics/mobius/blob/feature/ERC-223-support/contracts/Mixer.sol#L262). In fact, when the condition `if( ring.IsDead() )` is met, the entry `m_rings[ring_id];` is deleted, thus returning a storage pointer to this entry triggers an undesirable behavior in the calling function. This is why we needed to included a `memory` variable as copy of this entry. This variable is removed once the calling function ends its execution (see: https://solidity.readthedocs.io/en/latest/frequently-asked-questions.html#what-is-the-memory-keyword-what-does-it-do).

While adding this temporary `entrySaved` variable might not be optimal, this prevents from code duplication and makes possible to defer external calls to the end of the `DepositERC20Compatible` and `WithdrawERC20Compatible` functions.

**Comment by AntoineRondelet** [comment]
Some solidity syntax best practices might need to be added to Mobius, however, this should be part of another PR.
I'm happy to do it, but since this PR includes breaking changes, I'd be happy to have a review @shapeshed and @zoenolan, so we can merge it if all is good for you.

File: README.md [hunk 1]
```diff
@@ -21,6 +21,14 @@ Möbius is a Smart Contract that runs on Ethereum that offers trustless autonomo
 
 :point_right: **See the full [Tutorial](./mobiusTutorial.md)**
 
+-----------------------------------------------------
+
+Möbius supports ether transactions and ERC20 compatible token transactions.
```
Comment (zoenolan) [inline README.md]: I would delete the first 'transactions'

File: contracts/Mixer.sol [hunk 2]
```diff
@@ -215,8 +274,12 @@ contract Mixer
 
         MixerWithdraw(ring_id, tag_x, entry.token, entry.denomination);
 
-        // TODO: add ERC-223 support
-        msg.sender.transfer(entry.denomination);
+        // We want to return a copy of the entry in order to be able to access
+        // The token and denomination fileds of this object.
```
Comment (zoenolan) [inline contracts/Mixer.sol]: No captial on 'The' and 'fileds' should be 'fields'

File: mobiusTutorial.md [hunk 3]
```diff
@@ -31,6 +31,16 @@ v9.5.0
 
 ## How to use Mobius ?
 
+---------------------------------------------------
+
+As Möbius supports transfer of ethers and ERC20 compatible tokens, we denote by `Deposit` and `Withdraw` - in this tutorial - the action to do a deposit and a withdrawal to and from the Mixer. The user is free to use the appropriate suffix `Ether` or `ERC20Compatible`.
+
+:warning: In order for the `DepositERC20Compatible` function to run, the sender has to authorize the `Mixer` to trigger a transfer of tokens on his behalf, **FIRST**. To do this, the sender has to run: `token.approve([mixerAddress], [amountToApprove], {from: [senderAccount]});` (see: ERC20 interface for more details), before calling `DepositERC20Compatible`.
```
Comment (zoenolan) [inline mobiusTutorial.md]: 'on their behalf'

File: README.md [hunk 4]
```diff
@@ -21,6 +21,14 @@ Möbius is a Smart Contract that runs on Ethereum that offers trustless autonomo
...
+In order to use the Mixer with ERC20 compatible tokens, one has to use the `DepositERC20Compatible` and `WithdrawERC20Compatible` functions.
```
Comment (zoenolan) [inline README.md]: I would use "`DepositERC20Compatible` and `WithdrawERC20Compatible` functions must be used"

File: contracts/ERC223ReceivingContract.sol [hunk 5]
```diff
@@ -0,0 +1,28 @@
+// Copyright (c) 2016-2018 Clearmatics Technologies Ltd
+
+// SPDX-License-Identifier: LGPL-3.0+
+
+pragma solidity ^0.4.19;
+
+/*
+ * Contract that is working with ERC223 tokens
```
Comment (zoenolan) [inline contracts/ERC223ReceivingContract.sol]: I would use 'ERC223 token compatible contract'

**Pull request merged**